webui/react/src/pages/F_ExpList/glide-table/ColumnPickerMenu.tsx

```diff
@@ -0,0 +1,234 @@
+import Input, { InputRef } from 'components/kit/Input';
+import Pivot from 'components/kit/Pivot';
+import { V1LocationType } from 'services/api-ts-sdk';
+import Spinner from 'shared/components/Spinner';
+import { isEqual } from 'shared/utils/data';
+import { ProjectColumn } from 'types';
+import { Loadable } from 'utils/loadable';
+
+import { defaultExperimentColumns } from './columns';
+
+const BANNED_COLUMNS = new Set(['name']);
+
+interface Props {
+  initialVisibleColumns: string[];
+  projectColumns: Loadable<ProjectColumn[]>;
+  setVisibleColumns: (newColumns: string[]) => void;
+}
+
+const ColumnPickerMenu: React.FC<Props> = ({
+  projectColumns,
+  setVisibleColumns,
+  initialVisibleColumns,
+}) => {
+  const [form] = Form.useForm();
+  const removeBannedColumns = useCallback(
+    (columns: ProjectColumn[]) => columns.filter((col) => !BANNED_COLUMNS.has(col.column)),
+    [],
+  );
+  const [totalColumns, setTotalColumns] = useState<ProjectColumn[]>(() =>
+    removeBannedColumns(Loadable.getOrElse([], projectColumns)),
+  );
+  const [filteredColumns, setFilteredColumns] = useState<Set<string>>(
+    () =>
+      new Set(removeBannedColumns(Loadable.getOrElse([], projectColumns)).map((col) => col.column)),
+  );
+  const [isColumnsOpen, setIsColumnsOpen] = useState(false);
+  const [activeColumnTab, setActiveColumnTab] = useState<V1LocationType>(V1LocationType.EXPERIMENT);
+  const searchRef = useRef<InputRef>(null);
+
+  const columnSearch: string = Form.useWatch('column-search', form) ?? '';
+
+  useEffect(() => {
+    setTotalColumns(removeBannedColumns(Loadable.getOrElse([], projectColumns)));
+  }, [projectColumns, removeBannedColumns]);
+
+  useEffect(() => {
+    const regex = new RegExp(columnSearch, 'i');
+    setFilteredColumns(
+      new Set(
+        totalColumns
+          .filter((col) => regex.test(col.displayName || col.column))
+          .map((col) => col.column),
+      ),
+    );
+  }, [columnSearch, removeBannedColumns, totalColumns]);
+
+  const generalColumns: Record<string, boolean> = Form.useWatch(V1LocationType.EXPERIMENT, form);
+  const hyperparametersColumns: Record<string, boolean> = Form.useWatch(
+    V1LocationType.HYPERPARAMETERS,
+    form,
+  );
+  const metricsColumns: Record<string, boolean> = Form.useWatch(V1LocationType.VALIDATIONS, form);
+
+  const allFormColumns = useMemo(
+    () => ({ ...generalColumns, ...hyperparametersColumns, ...metricsColumns }),
+    [generalColumns, hyperparametersColumns, metricsColumns],
+  );
+
+  useEffect(() => {
+    if (Object.keys(allFormColumns).length === 0) return;
+    /* eslint-disable @typescript-eslint/no-unused-vars */
+    setVisibleColumns(
+      Object.entries(allFormColumns)
+        .filter(([_, checked]) => checked)
+        .map(([column, _]) => column),
+    );
+    /* eslint-enable @typescript-eslint/no-unused-vars */
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [allFormColumns, setVisibleColumns]);
```

Comment:
We shouldn't be setting state in effects -- it causes extra renders that aren't necessary. Consider moving this functionality to an `onValuesChanged` handler on the `Form` element or not using the form handler entirely.